### PR TITLE
feat: adapt structure to template

### PR DESCRIPTION
**Pull request by jakob1111996**
Finished v1 of the BiG Transfer Encoder.
Changes:
- Added integration tests and updated tests
- CI Pipeline
- Download model directly in Python if model_path is empty

### CODE REVIEW

File: big_transfer.py [hunk 1]
```diff
@@ -20,61 +20,98 @@ class BigTransferEncoder(Executor):
...
-                f'model at {self.model_path} does not exist')
+        if not os.path.exists(self.model_path):
+            self.download_model()
+
+        cpus = tf.config.experimental.list_physical_devices(
+            device_type='CPU')
+        gpus = tf.config.experimental.list_physical_devices(
+            device_type='GPU')
+        if self.on_gpu and len(gpus) > 0:
+            cpus.append(gpus[0])
```
Comment (maximilianwerk) [inline big_transfer.py:75]: Not sure, but should we raise an Exception, if `self.on_gpu and len(gpus) == 0`?

**Review by maximilianwerk — COMMENTED** [review]
Looks good to me.

File: big_transfer.py (hunk 1 above)
Comment (jakob1111996) [reply big_transfer.py:75]: I did a warning here: https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/0f11e821d6f2ef4b8b690a273a8c0aae0ebfaedc

File: Dockerfile [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
+WORKDIR ./big_transfer
 
-RUN pip install -r requirements.txt
+RUN pip install .
 
-ENTRYPOINT ["jina", "pod", "--uses", "config.yml"]
\ No newline at end of file
+FROM base
+RUN pip install -r tests/requirements.txt
+RUN pytest tests
```
Comment (cristianmtr) [inline Dockerfile:10]:
```suggestion
RUN pytest -s -v tests
```

for verbose output in CI

File: Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+RUN apt update && apt install -y python3.8 python3-pip
+
+COPY . ./big_transfer/
+WORKDIR ./big_transfer
+
+RUN pip install .
+
+FROM base
+RUN pip install -r tests/requirements.txt
+RUN pytest tests
```
Comment (cristianmtr) [inline Dockerfile.gpu:12]: same as above

File: tests/unit/test_big_transfer.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    img = Image.fromarray(doc.blob.astype('uint8'))
+    img = img.resize((96, 96))
+    img = np.array(img).astype('float32') / 255
+    doc.blob = img
+    assert doc.embedding is None
+
+    encoder = BigTransferEncoder()
+
+    encoded_doc = encoder.encode(DocumentArray([doc]))
+    assert encoded_doc[0].embedding is not None
```
Comment (cristianmtr) [inline tests/unit/test_big_transfer.py:44]: assert an expected shape?

File: README.md [hunk 5]
```diff
@@ -13,5 +13,8 @@ It uses a pretrained version of a BiT model to encode an image from an array of
 The following parameters can be used:
 
 - `model_path` (string, default: "pretrained"): The folder where the downloaded pretrained Big Transfer model is located
+- `model_name` (string, default: "R50x1"): The model to be downloaded when the model_path is empty. Choose from ['R50x1', 'R101x1', 'R50x3', 'R101x3', 'R152x4']
```
Comment (Yongxuanzhang) [inline README.md:16]: For the README, we could refer to the template to have the same contents structure

File: big_transfer.py [hunk 6]
```diff
@@ -2,12 +2,12 @@
 __license__ = "Apache-2.0"
 
 import os
+from typing import Optional
 import numpy as np
 import tensorflow as tf
-from typing import Optional
+from tensorflow.python.keras.models import load_model
```
Comment (Yongxuanzhang) [inline big_transfer.py:8]: optimize the imports

File: big_transfer.py [hunk 7]
```diff
@@ -94,4 +134,4 @@ def encode(self, docs: DocumentArray, **kwargs) -> DocumentArray:
         output = _output['output_1'].numpy()
         for index, doc in enumerate(docs):
             doc.embedding = output[index]
-        return docs
\ No newline at end of file
+        return docs
```
Comment (Yongxuanzhang) [inline big_transfer.py:137]: do we have traversal path for this encoder?

File: README.md (hunk 5 above)
Comment (jakob1111996) [reply README.md:10]: https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/b845ac37e6f21776c11dbcfe17be6573e30e0208

File: tests/unit/test_big_transfer.py (hunk 4 above)
Comment (jakob1111996) [reply tests/unit/test_big_transfer.py]: Done https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/d1470567f411afdcbb68b5e4294af46cd3d8b527

File: Dockerfile.gpu (hunk 3 above)
Comment (jakob1111996) [reply Dockerfile.gpu]: https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/d1470567f411afdcbb68b5e4294af46cd3d8b527

File: Dockerfile (hunk 2 above)
Comment (jakob1111996) [reply Dockerfile]: https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/d1470567f411afdcbb68b5e4294af46cd3d8b527

File: big_transfer.py (hunk 6 above)
Comment (jakob1111996) [reply big_transfer.py]: What exactly do you mean here?

File: big_transfer.py (hunk 7 above)
Comment (jakob1111996) [reply big_transfer.py]: Implemented it just now: https://github.com/jina-ai/executor-big-transfer-encoder/pull/1/commits/0a03a3dc9d13f41ed3093dc6d8e1be64e4c60586

File: big_transfer.py [hunk 8]
```diff
@@ -20,78 +20,146 @@ class BigTransferEncoder(Executor):
...
+    :param default_traversal_paths: Traversal path through the docs
+    :param default_batch_size: Batch size to be used in the encoder model
 
-        :param channel_axis: the axis id of the channel, -1 indicate the color
-            channel info at the last axis. If given other, then `
-            `np.moveaxis(data, channel_axis, -1)`` is performed before :meth:`encode`.
+    :param: on_gpu: If true, the GPU will be used. Make sure to have
+        tensorflow-gpu==2.5 installed
     """
+    DEFAULT_TRAVERSAL_PATHS = ['r']
```
Comment (Yongxuanzhang) [inline big_transfer.py:57]: I think we can put this directly as default param?
`default_traversal_paths: List[str] =['r'],`

**Review by Yongxuanzhang — APPROVED** [review]

**Pull request merged**